### ISSUE TITLE
Support Variable Interest Rates

### ISSUE THREAD

**Issue by stellarpower** [ArtyomPanfutov/loan-amortization-calculator#40]
Hi,

Been experimenting with an interest rate that isn't static - so far, looking at the docs, it seems that the library is mostly built around a rate fixed for the lifetime of the loan. I see in some other countries it seems a 30-year fix is possible, but where I live there is no chance of fixing for more than 5 years, and many products are trackers or variable. A common product is also an initial fixed rate, which then reverts to a tracker/SVR after the fix ends in 1/2/3/5 years, so I have so far managed to simulate this situation by creating two separate amortisation schedules - an initial one, for the fixed rate, and then one at a higher (but still assumed invariable, for now) rate for the rest of the product life, without having to break into the builder classes. However, it would be a nice feature if we could pass in an iterable of different interest rates for each month, or perhaps take some type of generator object, to forecast changes in the base rates and to simulate variable mortgages.

Thanks for the nice work BTW. I'm connecting this straight into KNIME, (eventually) mangling the data tables in there and calling out to the JAR - sooo much better than any more awful spreadsheets :D

**Comment by ArtyomPanfutov** [comment]
Hello,

Sorry for the long response. 

You are right - I didn't think about variable interest rates while developing this lib. 
I like the idea of some objects with custom implementation for providing interest rates for each month. 

I'll try to show you the experimental implementation in the near future and we can discuss it. 

Thank you for your request!

**Comment by stellarpower** [comment]
Sounds good! More than happy to give a try and provide feedback - although I'm not a Java developer just FYI, so that will limit how helpful I can be. In C++ I think using generics with something overloading operator() is a nice way of doing it, so I like what you're describing. Could then maybe provide a couple of somple classes that implement this interface and provide for typical patterns using just a few args in the ctor - stepped rates, linearly-increasing, feeding items from an array, etc. etc.

Cheers :D

**Comment by stellarpower** [comment]
I'm going to be looking at all this (simulating real rates, that is) this weekend, just if you happen to have anything prepared! Should I check out (and checkout) this branch: feature/custom-interest-rate ?

**Comment by ArtyomPanfutov** [comment]
Yes, you can find the changes in this branch. Make sure you have the updated version because my last push was several minutes ago. 

I've added a parameter in the loan class which is an entry point for providing custom monthly interest rates. It is a functional interface with input and output.
This function will be called from the calculator implementation for every monthly payment
The input is additional data that is needed for a calculation. For now, I've added only the payment number to this class. 
The must be an interest rate for the requested month.

Usage example you can find in LoanAmortizationCalculatorTest#shouldUseCustomInterestRate 
``` Java
   @Test
    void shouldUseCustomInterestRate() {
        final List<BigDecimal> rates = Arrays.asList(
                BigDecimal.valueOf(0.0036),
                BigDecimal.valueOf(0.0031),
                BigDecimal.valueOf(0.0031),
                BigDecimal.valueOf(0.0053),
                BigDecimal.valueOf(0.0022));

        final Loan loan = Loan.builder()
                .amount(BigDecimal.valueOf(500000.32))
                .term(5)
                .monthlyInterestRateProvider(input -> rates.get(input.getPaymentNumber()))
                .build();

        LoanAmortization amortization = calculator.calculate(loan);
        assertNotNull(amortization);

        for (MonthlyPayment payment : amortization.getMonthlyPayments()) {
            assertEquals(rates.get(payment.getMonthNumber()), payment.getInterestRate());
        }
    }
``` 

It is a concept and I didn't check the correctness of calculations. So if you have real examples of such loan schedules I would be happy to use them in test data.

**Comment by stellarpower** [comment]
Okay, I have been fighting Java today, so will probably pull tomorrow morning. I have done some basic comparisions myself to test, but I don't have any accurate schedules to compare against so far, just looking at general numbers (introductory payments, follow-on payments, ..). Harder to find these as a consumer - do you know anywhere I can get/make some? If not, I think I have some spreadsheets that should generate full tables. I may as well also make a repo for my KNIME workflow and binding code, so you could use that for testing if you wanted at all.
